Automatic traits Events wrapper

vidartf opened the pull request:
By replacing `traits.api.HasTraits` with `hyperspy.Events.HasEventsTraits`, `traits` will work as previously, but an `Event` will be created for the change notifier of each trait. So if the class has traits A, B and C, this will automatically create `events.A_changed`, `events.B_changed` and `events.C_changed`. It also retains the callback argument signatures of traits (new; name, new; obj, name, new; obj, name, old, new).

vidartf committed:
Separate out Event nargs resolution

Make it a separate function to ease overloading.

vidartf committed:
Add HasEventsTraits class + tests

Any class that inherit HasEventsTraits, will have any detected traits
wrapped in an Event. The event will be called
`<Object>.<trait_name>_changed`. Currently only handles traits that are
defined statically, i.e. currently does nothing for traits added with
`add_trait()`/`add_class_trait()` which is important for components<->
parameters.

francisco-dlp commented:
This should make the transition to events much more smoother and it is also very cool. However, I was dreaming with getting rid of traits at some point and this PR pushes in the opposite direction. I bet that you think that getting rid of traits is not a good idea, could you share your reasoning with us?

vidartf commented:
I think the first step in getting rid of traits is to insulate ourselves from it. The next step here would be to add a similar class for traitlets, and make it work the same from the users point of view (possibly also moving away from the silly args resolution of traits). Then, the traits wrapper can be removed if/when the transition is completed.

vidartf commented:
In other words: Making the connecting points from our code to traits as few as possible should ease the transition later. This PR does that, and it makes internal hyperspy code traits-package agnostic.

francisco-dlp commented:
Ok, I get you point although I'm not yet 100% sure that this is the way to
go. All is nicely done and in principle ready to merge. However, I propose
to delay merging this until we make further progress in transitioning to
our own events. I think that then we'll be able to better evaluate the
advantages and disadvantages of this approach.

2015-12-11 19:55 GMT+00:00 Vidar Tonaas Fauske notifications@github.com:

> In other words: Making the connecting points from our code to traits as
> few as possible should ease the transition later. This PR does that, and it
> makes internal hyperspy code traits-package agnostic.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/francisco-dlp/hyperspy/pull/22#issuecomment-164033778
> .

vidartf commented:
I get that you're not a 100%, but delaying the decision will only make it harder: There are in principle two things that need to be done to transition to our own events:
1. Create a new event. This means giving it s name, and defining/documenting when and where it triggers.
2. Have existing code connect to the Event and use it.

This PR basically automates the first step for all change events of traits. If we start the transition manually, we might end up introducing patterns that are incompatible with what is proposed here, making this PR obsolete.

That being said, I'm ready to forgo the "traits compatible nargs resolution", but the automatic creation and naming of events is really useful. We wouldn't have to replace all instance of `HasTraits` with `HasEventsTraits` either, only those where we want these automatic events.

francisco-dlp commented:
The thing is that we don't depend on traits that much. Basically, the only
core feature that depends on traits is the `AxesManager` and there traits
is easy and advantageous to replace. In all other cases, we use traits only
for the convenience of  easily creating UI elements using traitsui and we
actually don't need events in those case for anything but the UI. But
wouldn't it be good to move all the UI elements somewhere else
(HyperSpyUI?) and replace them with ipywidgets in HyperSpy?

2015-12-17 14:04 GMT+00:00 Vidar Tonaas Fauske notifications@github.com:

> I get that you're not a 100%, but delaying the decision will only make it
> harder: There are in principle two things that need to be done to
> transition to our own events:
> 1. Create a new event. This means giving it s name, and
>    defining/documenting when and where it triggers.
> 2. Have existing code connect to the Event and use it.
> 
> This PR basically automates the first step for all change events of
> traits. If we start the transition manually, we might end up introducing
> patterns that are incompatible with what is proposed here, making this PR
> obsolete.
> 
> That being said, I'm ready to forgo the "traits compatible nargs
> resolution", but the automatic creation and naming of events is really
> useful. We wouldn't have to replace all instance of HasTraits with
> HasEventsTraits either, only those where we want these automatic events.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/francisco-dlp/hyperspy/pull/22#issuecomment-165460415
> .

vidartf commented:
Even if we're not replacing existing events structures, it is useful to have a common way of creating events for things that (currently) use traits, e.g. `Parameters`. Hopefully, we will also make hyperspy's interactive plots more event based, whether they use traits or traitlets.

vidartf commented:
I'm currently trying to replace the code in `DataAxis` to use Events, and I'm really struggling to find a way to reliably wire traits to events. The issue is with the silly resolution of args that traits has. See current code below:

``` python
# ... Added to end of DataAxis.__init__() instead of current self.on_trait_change code
        self.events.scale_changed = Event()
        self.events.offset_changed = Event()
        self.events.value_changed = Event()
        self.events.index_changed = Event()
        self.events.size_changed = Event()
        self.events.navigate_changed = Event()
        self.events.scale_changed.connect(self.update_axis, 0)
        self.events.offset_changed.connect(self.update_axis, 0)
        self.events.size_changed.connect(self.update_axis, 0)
        self.events.index_changed.connect(self.update_value, 0)
        self.events.value_changed.connect(self.set_index_from_value, 1)
        self.events.size_changed.connect(self.update_index_bounds, 0)
        self.events.navigate_changed.connect(self._update_slice, 1)
        for tn in ['scale', 'offset', 'size', 'index', 'value',
                   'navigate', 'size']:
            e = getattr(self.events, tn + '_changed')
            f = e.trigger       # This won't work, as we never get any args!
            self.on_trait_change(f, tn)
```

For these default events, `set_index_from_value()` and `_update_slice` will never get the argument they're expecting, since the Event trigger will not receive any arguments. We can create any variation of different types of enclosures to work around this problem, but the most elegant solution by far is the code that is in this PR...

vidartf commented:
PS: I force pushed a rebased code base.

vidartf commented:
PPS: Alternative solutions are of course welcome :)

vidartf committed:
Avoid metaclass!

This was simpler than it first appeared. The only requirement is that
the class calls the init of super at the right spot...

vidartf commented:
I have now also removed the metaclass code. I figured out it just made things more complicated. Now there's only a requirement that the init function is called before trying to access trait related events.

francisco-dlp commented:
@vidartf, I agree that this PR is a very elegant way of getting hspy `Events` "for free" from traits. However, ideally what we want is to replace traits, not to provide an alternative way to connect to their events. I see that using the solution that you propose here could make the transition smoother once we replace traits in this code with our own events. However, why not doing it now? The only benefit of traits here is the automatic generation of the sliders (or am I missing something?), but it shouldn't be hard to implement the sliders with wxPython and PyQt directly, isn't it?

vidartf commented:
@francisco-dlp : Maybe I misunderstood something. In hyperspy/hyperspy#423 there is a todo point called "Reimplement old events using the new infraestructure.". Should this also include removing `Traits` from `DataAxis`/`AxesManager`? If so, then this PR is not needed, but it will also be quite a lot of work to get events merged... If that is not the case, then in the interim we will need some way to connect `Traits` to our events (if we replace traits with traitlets we will also need an interface then). If you have an alternative way of doing that other than this PR, then I'm happy to hear it, but this is my best attempt.

francisco-dlp commented:
What I had in mind there was mainly replacing all the events implemented by keeping a `connected` list in the object. So, no, I didn't mean "replace traits with events" when I wrote that, sorry for the confusion.
